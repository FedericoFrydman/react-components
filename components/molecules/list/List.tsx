import { Ref, forwardRef, ReactElement } from 'react';

import './list.scss';
import { Container, ContainerProps } from '../../../components/atoms/container/Container';

const classList = 'lsux-list';

export interface ListProps extends ContainerProps {
  /**
   * The children of the list.
   */
  children?: string | number | ReactElement | ReactElement[];
  /**
   * Determines if each list item has a border.
   * @default false
   */
  noBorder?: boolean;
  /**
   * Stretch the component across the full width of its parent.
   */
  stretch?: boolean;
  /**
   * Additional classNames to add to the list.
   */
  classNames?: string[];
  /**
   * The ref to the HTML DOM element.
   */
  ref?: Ref<HTMLDivElement>;
}

const ListComponent = (
  { children, noBorder = false, stretch, classNames = [], ...props }: ListProps,
  ref: ListProps['ref'],
) => {
  const cnList = {
    [`${classList}`]: true,
    [`${classList}--border`]: !noBorder,
    [`${classList}--stretch`]: stretch,
  };

  return (
    <Container {...props} ref={ref} classNames={[cnList, ...classNames]}>
      {children}
    </Container>
  );
};

export const List = forwardRef<HTMLDivElement, ListProps>(ListComponent);

/**
 * The List component displays content in additional sub-components for content, inputs, and actions.
 */
