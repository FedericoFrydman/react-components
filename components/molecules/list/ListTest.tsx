import { forwardRef, useState } from 'react';
import { Card, Container, List, ListItemInput, ListProps, Text } from '../../../';

const ListTestComponent = ({ ...props }: ListProps) => {
  const exampleItems = ['0', '1', '2', '3', '4', '5', '6', '7', '8', '9'];
  const [selectedIds, setSelectedIds] = useState(['1']);

  function handleSelectionChange(id: string) {
    if (selectedIds.includes(id)) {
      setSelectedIds(selectedIds.filter((i) => i !== id));
    } else {
      setSelectedIds([...selectedIds, id]);
    }
  }

  const checkboxItems = exampleItems.map((id, i) => (
    <ListItemInput.Switch
      key={id}
      id={id.toString()}
      inputProps={{ name: id.toString() }}
      primary={`$${i + 1}0.00/mo`}
      secondary="Lorem ipsum dolor sit amet"
      isSelected={selectedIds.includes(id)}
      handleClick={handleSelectionChange}
    />
  ));

  return (
    <Container classNames={['m-5']} style={{ maxWidth: 340 }}>
      <Card background="white">
        <Card.Content background="white">
          <Card.Title>Plan Name</Card.Title>
          <Text text="Example plan description text" />
        </Card.Content>
        <Card.Content background="light-gray">
          <Card.Title classNames={['mb-3']}>Coverage Add-ons</Card.Title>
          <List background="white" {...props}>
            {checkboxItems}
          </List>
        </Card.Content>
      </Card>
    </Container>
  );
};

export const ListTest = forwardRef<HTMLDivElement, ListProps>(ListTestComponent);
