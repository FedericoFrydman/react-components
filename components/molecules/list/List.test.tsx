import { shallow, ShallowWrapper } from 'enzyme';
import { render, screen } from '@testing-library/react';
import cn from 'classnames';

import { ListProps } from './List';
import { ListTest } from './ListTest';

describe('List Component', () => {
  let testId: string;
  let wrapper: ShallowWrapper;

  beforeEach(() => {
    testId = 'test list';
    wrapper = shallow(<ListTest />);
  });

  describe('props', () => {
    describe('classNames', () => {
      describe('when provided classNames', () => {
        const classNames = ['my-class-1', 'my-class-2'];

        it('gives the list the provided classNames', () => {
          const list = renderList({ classNames });

          expect(list).toHaveClass(cn(classNames), { exact: false });
        });
      });
    });

    describe('noBorder', () => {
      it('should contain border class when false', () => {
        const list = renderList();
        expect(list).toHaveClass('lsux-list--border');
      });
      it('should not contain border class when true', () => {
        const list = renderList({ noBorder: true });
        expect(list).not.toHaveClass('lsux-list--border');
      });
    });
  });

  test('should render the component', () => {
    expect(wrapper).toMatchSnapshot();
  });

  const renderList = (props: ListProps = {}): HTMLElement => {
    render(<ListTest {...props} data-testid={testId} />);

    return screen.queryByTestId(testId);
  };
});
