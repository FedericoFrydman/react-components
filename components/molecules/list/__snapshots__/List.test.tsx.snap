// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`List Component should render the component 1`] = `
<ForwardRef(ContainerComponent)
  classNames={
    Array [
      "m-5",
    ]
  }
  style={
    Object {
      "maxWidth": 340,
    }
  }
>
  <Card
    background="white"
  >
    <CardContent
      background="white"
    >
      <CardTitle>
        Plan Name
      </CardTitle>
      <ForwardRef(TextComponent)
        text="Example plan description text"
      />
    </CardContent>
    <CardContent
      background="light-gray"
    >
      <CardTitle
        classNames={
          Array [
            "mb-3",
          ]
        }
      >
        Coverage Add-ons
      </CardTitle>
      <ForwardRef(ListComponent)
        background="white"
      >
        <ListItemInputSwitchComponent
          handleClick={[Function]}
          id="0"
          inputProps={
            Object {
              "name": "0",
            }
          }
          isSelected={false}
          key="0"
          primary="$10.00/mo"
          secondary="Lorem ipsum dolor sit amet"
        />
        <ListItemInputSwitchComponent
          handleClick={[Function]}
          id="1"
          inputProps={
            Object {
              "name": "1",
            }
          }
          isSelected={true}
          key="1"
          primary="$20.00/mo"
          secondary="Lorem ipsum dolor sit amet"
        />
        <ListItemInputSwitchComponent
          handleClick={[Function]}
          id="2"
          inputProps={
            Object {
              "name": "2",
            }
          }
          isSelected={false}
          key="2"
          primary="$30.00/mo"
          secondary="Lorem ipsum dolor sit amet"
        />
        <ListItemInputSwitchComponent
          handleClick={[Function]}
          id="3"
          inputProps={
            Object {
              "name": "3",
            }
          }
          isSelected={false}
          key="3"
          primary="$40.00/mo"
          secondary="Lorem ipsum dolor sit amet"
        />
        <ListItemInputSwitchComponent
          handleClick={[Function]}
          id="4"
          inputProps={
            Object {
              "name": "4",
            }
          }
          isSelected={false}
          key="4"
          primary="$50.00/mo"
          secondary="Lorem ipsum dolor sit amet"
        />
        <ListItemInputSwitchComponent
          handleClick={[Function]}
          id="5"
          inputProps={
            Object {
              "name": "5",
            }
          }
          isSelected={false}
          key="5"
          primary="$60.00/mo"
          secondary="Lorem ipsum dolor sit amet"
        />
        <ListItemInputSwitchComponent
          handleClick={[Function]}
          id="6"
          inputProps={
            Object {
              "name": "6",
            }
          }
          isSelected={false}
          key="6"
          primary="$70.00/mo"
          secondary="Lorem ipsum dolor sit amet"
        />
        <ListItemInputSwitchComponent
          handleClick={[Function]}
          id="7"
          inputProps={
            Object {
              "name": "7",
            }
          }
          isSelected={false}
          key="7"
          primary="$80.00/mo"
          secondary="Lorem ipsum dolor sit amet"
        />
        <ListItemInputSwitchComponent
          handleClick={[Function]}
          id="8"
          inputProps={
            Object {
              "name": "8",
            }
          }
          isSelected={false}
          key="8"
          primary="$90.00/mo"
          secondary="Lorem ipsum dolor sit amet"
        />
        <ListItemInputSwitchComponent
          handleClick={[Function]}
          id="9"
          inputProps={
            Object {
              "name": "9",
            }
          }
          isSelected={false}
          key="9"
          primary="$100.00/mo"
          secondary="Lorem ipsum dolor sit amet"
        />
      </ForwardRef(ListComponent)>
    </CardContent>
  </Card>
</ForwardRef(ContainerComponent)>
`;
